src/backend/backend/tasks.py

```diff
@@ -0,0 +1,124 @@
+import asyncio
+import io
+import json
+import uuid
+from pathlib import Path
+
+import httpx
+from a2a.types import JSONRPCErrorResponse, SendMessageSuccessResponse
+
+from backend.celery_app import celery_app
+from backend.client import AgentClient
+from backend.db import session_manager
+from backend.logging import logger
+from backend.repositories.agents import AgentRepository
+from backend.schemas import AgentStatus
+from backend.settings import settings
+from backend.storage import storage_client
+from backend.templates import AGENT_CODE_TEMPLATE, DOCKERFILE_TEMPLATE
+
+BASE_URL = settings.A2A_AGENT_URL
+
+
+def _get_agent_repository():
+    with session_manager.session() as session:
+        yield AgentRepository(session)
+
+
+def _update_agent_status(agent_id: uuid.UUID, status: str):
+    logger.info(f"Updating agent {agent_id} status to {status}")
+    for repo in _get_agent_repository():
+        repo.update(agent_id, status=status)
+
+
+def _save_agent(result: dict[str, str], agent_id: uuid.UUID) -> None:
+    bucket_name = settings.S3_BUCKET_NAME
+    logger.info(f"Saving agent files to bucket {bucket_name} for agent {agent_id}...")
+    try:
+        storage_client.head_bucket(Bucket=bucket_name)
+    except Exception:
+        storage_client.create_bucket(Bucket=bucket_name)
+
+    try:
+        agent_code = AGENT_CODE_TEMPLATE.format(**result)
+        storage_client.put_object(
+            Bucket=bucket_name,
+            Key=f"{agent_id}/agent.py",
+            Body=io.BytesIO(agent_code.encode()),
+            ContentType="text/x-python",
+        )
+
+        storage_client.put_object(
+            Bucket=bucket_name,
+            Key=f"{agent_id}/README.md",
+            Body=io.BytesIO(result["readme"].encode()),
+            ContentType="text/markdown",
+        )
+
+        storage_client.put_object(
+            Bucket=bucket_name,
+            Key=f"{agent_id}/requirements.txt",
+            Body=io.BytesIO(result["dependencies"].encode()),
+            ContentType="text/plain",
+        )
+
+        storage_client.put_object(
+            Bucket=bucket_name,
+            Key=f"{agent_id}/Dockerfile",
+            Body=io.BytesIO(DOCKERFILE_TEMPLATE.encode()),
+            ContentType="text/plain",
+        )
+
+        tools_dir = Path("backend/tools")
+        for tool_file in tools_dir.iterdir():
+            if tool_file.is_file() and (tool_file.stem in agent_code or tool_file.name == "__init__.py"):
+                storage_client.upload_file(
+                    str(tool_file),
+                    bucket_name,
+                    f"{agent_id}/tools/{tool_file.name}",
+                )
+
+        logger.info(f"Agent files saved to bucket {bucket_name}")
```

Comment:
We weren't ever updating the status of the download to mark it as available.

```suggestion
        logger.info(f"Agent files saved to bucket {bucket_name}")
        _update_agent_download(agent_id, True)
```